source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -38,19 +39,50 @@ class CacheFilter : public Http::PassThroughFilter,
   Http::FilterDataStatus encodeData(Buffer::Instance& buffer, bool end_stream) override;
 
 private:
+  // Utility functions; make any necessary checks and call the corresponding lookup_ functions
   void getBody();
-  void onHeaders(LookupResult&& result);
+  void getTrailers();
+
+  // Callbacks for HttpCache to call when headers/body/trailers are ready
+  void onHeaders(LookupResult&& result, Http::RequestHeaderMap& request_headers);
   void onBody(Buffer::InstancePtr&& body);
   void onTrailers(Http::ResponseTrailerMapPtr&& trailers);
 
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  //               filter_state_ is ValidatingCachedResponse
+  // Serves a validated cached response after updating it with a 304 response
+  void processSuccessfulValidation(Http::ResponseHeaderMap& response_headers);
+
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  //               filter_state_ is ValidatingCachedResponse
+  // Checks if a cached entry should be updated with a 304 response
+  bool shouldUpdateCachedEntry(const Http::ResponseHeaderMap& response_headers) const;
+
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  // Should only be called during onHeaders as it modifies RequestHeaderMap
+  // Adds required conditional headers for cache validation to the request headers
+  // according to the present cache lookup result headers
+  void injectValidationHeaders(Http::RequestHeaderMap& request_headers);
+
+  // Precondition: lookup_result_ points to a fresh or validated cache look up result
+  //               filter_state_ is ValidatingCachedResponse
+  // Adds a cache lookup result to the response encoding stream
+  // Can be called during decoding if a valid cache hit is found
+  // or during encoding if a cache entry was validated successfully
+  void encodeCachedResponse();
+
+  // Precondition: finished adding a response from cache to the response encoding stream
+  // Updates the filter_state_ and continues the encoding stream if necessary
+  void finishedEncodingCachedResponse();
```

Comment:
This function should be called after the cache body is encoded to the stream successfully. It doesn't actually "finish" the encoding, it just updates the state and calls `encoder_callbacks_->continueEncoding()` if it had been stopped (if `encodeHeaders` returned `StopIteration`). I am not fond of that name either but I think `finish` could be confusing? What do you think?

Reply:
continueFinishedResponse?

However, I think this naming difficulty is a sign that something needs to be restructured. In this case, I think inlining finishedEncodingCachedResponse into the two functions that call it will make the overall code structure easier to understand. In particular, it will make it easy to see that onTrailers moves us to the ResponseServedFromCache state, and that onBody might.